### PR TITLE
Make parsing independent of the cumulativity flag.

### PR DESCRIPTION
**Pull request by SkySkimmer**
This fixes the issue mentioned in https://github.com/coq/coq/issues/7015#issuecomment-374046176 without messing with the STM and is cleaner.
NB: I don't know how to add it to the test suite.

### CODE REVIEW

**Comment by gares** [comment]
The STM needs to know which sentences alter the parser.
The Set Polymorphic Inductive Cumulativity one does:
https://github.com/coq/coq/blob/master/parsing/g_vernac.ml4#L173
but is not classified as such (VtNow)
https://github.com/coq/coq/blob/master/stm/vernac_classifier.ml#L58

**Comment by gares** [comment]
This is likely to be fixed by the "attribute" system

**Comment by gares** [comment]
This PR fixes the issue by moving the use of the flag from parsing time to interpretation time

File: intf/vernacexpr.ml
Comment (vbgl) [inline intf/vernacexpr.ml]: Can you please also update the comment, and explain a bit the meanings of the three states of this “boolean”.

**Review by vbgl — CHANGES_REQUESTED** [review]

File: intf/vernacexpr.ml
Comment (SkySkimmer) [reply intf/vernacexpr.ml]: Comment updated.

**Comment by Zimmi48** [comment]
Some secondary code owners have been removed. You may dismiss their request for review.

**Review by vbgl — APPROVED** [review]

**Comment by gares** [comment]
> Some secondary code owners have been removed. You may dismiss their request for review.

It seems I cannot, and I'd love to (also because the changes in some components are trivial/obliged, no point in pinging the maintainers)

The only interesting bit left to be reviewed is the change in `intf/: @letouzey, @ppedrot are you happy with it?

IMO there is not way to fix this without doing a change there (one has at least to add an option type around the old type to signal "use the global flag, the user did not specify it") and this implies the backport breaks the API published in the beta. (you may want to change the milestone @Zimmi48)

**Review by gares — APPROVED** [review]

**Comment by SkySkimmer** [comment]
>the backport breaks the API published in the beta

Good catch, I didn't notice the implication.

File: vernac/vernacentries.ml
Comment (herbelin) [inline vernac/vernacentries.ml]: Is this the (only) "relevant" line which makes the PR fixes the issue? (Or do I miss the point?)

If so, let me hijack (again) a PR. I understood that there was some kind of consensus, as a documentation, towards using more types with expliclty named constructions rather than booleans, as well as using labels to document booleans governing optional behaviors (at least, this was what @maximedenes was recommending at some time, and this sounds relevant to me). So, what is the general opinion about maybe trying to preserve the names `LocalCumulativity`, etc., for easier understanding and reasoning about the code?

Comment (gares) [reply vernac/vernacentries.ml]: > Is this the (only) "relevant" line which makes the PR fixes the issue? (Or do I miss the point?)

Accessing Flags at parse time is not ok. The change to `bool option` is orthogonal.

Comment (SkySkimmer) [reply vernac/vernacentries.ml]: I wouldn't say that it's orthogonal, you have to put in something when there's no local flag.

**Comment by SkySkimmer** [comment]
Here's a commit which doesn't use `bool option`.

**Comment by gares** [comment]
Looks good to me now. I need an ack from:
- @letouzey  for intf
- @herbelin for parsing/printing
- @forestjulien for funind
I cover vernac.

**Comment by forestjulien** [comment]
No problem for me

**Review by forestjulien — APPROVED** [review]

**Comment by Zimmi48** [comment]
@forestjulien BTW not important at all but only one maintainer needs to self-assign (so as to indicate that they will merge the PR when it is ready). In this case, @gares had already done it.

**Comment by herbelin** [comment]
OK to me.

**Comment by gares** [comment]
Ping @letouzey || @ppedrot w.r.t. the changes in `intf/`

**Comment by gares** [comment]
Ping @letouzey || @ppedrot w.r.t. the changes in intf/

**Review by ppedrot — APPROVED** [review]
I don't think I have much to say here, but it looks OK to me anyways.

**Pull request merged**
